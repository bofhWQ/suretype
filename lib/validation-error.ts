import type * as Ajv from "ajv"
import { prettify } from "awesome-ajv-errors"


export interface ValidationErrorData
{
	errors: Array< Ajv.ErrorObject >;
	explanation?: string;
}
export interface ValidationResultInvalid extends ValidationErrorData
{
	ok: false;
}
export interface ValidationResultValid
{
	ok: true;
	errors?: Array< Ajv.ErrorObject >;
	explanation?: string;
}
export type ValidationResult = ValidationResultInvalid | ValidationResultValid;

export class ValidationError extends Error implements ValidationErrorData
{
	// @ts-ignore
	explanation: string;

	constructor(
		public errors: Array< Ajv.ErrorObject >,
		schema: unknown,
		data: unknown
	)
	{
		super( );

		makeExplanationGetter( this, 'explanation', errors, schema, data );
	}
}

export function makeExplanation(
	errors: Array< Ajv.ErrorObject >,
	schema: unknown,
	data: unknown,
	noFallback = false
)
{
	if ( schema && typeof schema === 'object' )
		return prettify( {
			errors,
			schema,
			data
		} );
	else if ( !noFallback )
		return JSON.stringify( errors, null, 2 );
	else
		return undefined;
}

export function makeExplanationGetter< T extends { }, P extends string >(
	target: T,
	property: P,
	errors: Array< Ajv.ErrorObject >,
	schema: unknown,
	data: unknown,
	noFallback = false
)
: T & { [ p in P ]: string; }
{
	let cache: string | undefined = undefined;
	Object.defineProperty( target, property, {
		get( )
		{
			if ( cache !== undefined )
				return cache;

			cache = makeExplanation( errors, schema, data, noFallback );
			return cache;
		}
	} );

	return target as any;
}
