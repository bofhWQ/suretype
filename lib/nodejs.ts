export * from './index'


// Patch the ValidationError to make Node.js console printing prettier

import { inspect } from "util"
import { ValidationError } from './validation-error'

( ValidationError.prototype as any )[ inspect.custom ] = function( )
{
	return ( this as ValidationError ).explanation;
}
