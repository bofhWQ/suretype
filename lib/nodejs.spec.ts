process.env.FORCE_COLOR = '0';

import { inspect } from "util"
import { v, compile } from './nodejs'
import { ValidationError } from './validation-error'


const getError = ( fn: ( ) => any ): Error =>
	{
		try
		{
			fn( );
		}
		catch ( err )
		{
			return err;
		}
		throw new EvalError( 'Function should have thrown' );
	};

describe( "nodejs", ( ) =>
{
	it( "should inject decorated console output for ValidationError", ( ) =>
	{
		expect( ( ValidationError.prototype as any )[ inspect.custom ] )
			.toStrictEqual( expect.any( Function ) );
	} );

	it( "should get decorated ValidationError", ( ) =>
	{
		const schema = v.object( { foo: v.number( ) } );
		const validator = compile( schema, { ensure: true } );
		const error = getError( ( ) => validator( { foo: "42" } ) );

		expect( inspect( error ) ).toMatchSnapshot( );
	} );
} );
