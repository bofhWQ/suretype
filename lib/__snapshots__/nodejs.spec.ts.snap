// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nodejs should get decorated ValidationError 1`] = `
"The type of the .foo value should be number, e.g. 42 (as number)

  1 | {
> 2 |   \\"foo\\": \\"42\\"
    |          ^^^^ Replace this with e.g. 42 (as number)
  3 | }"
`;
